src/main/scala-3/net/codingwell/scalaguice/TypeConversions.scala

```diff
@@ -0,0 +1,218 @@
+        //  TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),class Unit)
+        // )
+        // where "ByNameType" is shown as "ExprType" in debug output
+        case ByNameType(underlying: TypeRepr) => javaTypeExpr(underlying)
+
+        // For example, SomeClazz with Augmentation is AndType(
+        //   TypeRef(ThisType(TypeRef(NoPrefix,module class russ)),class SomeClazz),
+        //   TypeRef(ThisType(TypeRef(NoPrefix,module class russ)),trait Augmentation)
+        // )
+        case AndType(firstType: TypeRepr, _) => javaTypeExpr(firstType)
+
+        // Ignore annotations on types
+        case AnnotatedType(underlying, _) => javaTypeExpr(underlying)
+
+        // These generally happen when you bind[Something.type]
+        case ref: TermRef => toJavaType(ref)
+
+        // And then there's everything else that is possible, support through "best effort" translation
+        case x: ConstantType => bestEffort(x, "ConstantType")
+        case x: SuperType => bestEffort(x, "SuperType")
+        case x: Refinement => bestEffort(x, "Refinement")
+        case x: OrType => bestEffort(x, "OrType")
+        case x: MatchType => bestEffort(x, "MatchType")
+        case x: ParamRef => bestEffort(x, "ParamRef")
+        case x: ThisType => bestEffort(x, "ThisType")
+        case x: RecursiveThis => bestEffort(x, "RecursiveThis")
+        case x: RecursiveType => bestEffort(x, "RecursiveType")
+        case x: MethodType => bestEffort(x, "MethodType")
+        case x: PolyType => bestEffort(x, "PolyType")
+        case x: TypeLambda => bestEffort(x, "TypeLambda")
+        case x: MatchCase => bestEffort(x, "MatchCase")
+        case x: NoPrefix => bestEffort(x, "NoPrefix")
+
+        // And in case we missed something above
+        case x: TypeRepr => bestEffort(x, "TypeRepr (fell through)")
+      }
+
+    // Recursively evaluate and collect Scala type representations as higher-level Java type expressions
+    // Only handling the "and" case right (e.g., A with B with C)
+    def collectBoundExprs(scalaType: TypeRepr, exclude: Option[TypeRepr] = None): Seq[Expr[JavaType]] =
+      (scalaType.dealias, exclude) match {
+        case (ref: TypeRef, Some(excludedType)) if ref =:= excludedType =>
+          Seq.empty
+
+        case (ref: TypeRef, _) =>
+          Seq(javaTypeExpr(ref))
+
+        case (AndType(firstType, secondType), _) =>
+          collectBoundExprs(firstType, exclude) ++ collectBoundExprs(secondType, exclude)
+
+        case (other: TypeRepr, _) =>
+          Seq(bestEffort(other, "bound"))
+      }
+
+    // Anything we don't handle directly above is translated to a Java type using this default means
+    // (where logging might be nice to have while debugging)
+    def bestEffort(x: TypeRepr, tag: String = ""): Expr[JavaType] = {
+      report.warning(s"Encountered unexpected $tag type during macro expansion of ${x} -- using default translation")
```

Comment:
I'm not sure I would know what to do if I were to see this. Maybe you could include more details about where this is coming from, how to remediate, etc?

Reply:
Yeah, you are correct, it was a bit dodgy. Even if "best effort" might do the right thing some of the time, it was still kind of a bad idea. I've changed these all to just report various "unsupported" errors at compile time (often because Guice would arrive at a type that was too broad to be useful). It should be much clearer to the user why the error is happening and what to do about it now. I added some tests to demonstrate the desired error reporting in a few cases, but it isn't clear how to generate all these esoteric types that are theoretically possible. I believe what I've done should be good enough, and these could be revisited as time goes on.